File: codigo/matlab/newton_raphson.m
function [raiz, iter] = newton_raphson(f, df, x0, tol, maxit)
% NEWTON_RAPHSON  Metodo de Newton-Raphson. Chapra cap. 6; Sauer cap. 1.
%   f, df : handles de la funcion y su derivada
%   Uso:  [r,n] = newton_raphson(@(x) x.^3-x-2, @(x) 3*x.^2-1, 1.5, 1e-8, 50);
    x = x0;
    for iter = 1:maxit
        dfx = df(x);
        if dfx == 0, error('Derivada nula; el metodo falla.'); end
        x_new = x - f(x)/dfx;
        if abs(x_new - x) < tol, x = x_new; break; end
        x = x_new;
    end
    raiz = x;
end

File: codigo/python/newton_raphson.py
"""newton_raphson.py -- Metodo de Newton-Raphson. Chapra cap. 6; Sauer cap. 1."""


def newton_raphson(f, df, x0, tol=1e-8, maxit=50):
    x = x0
    for i in range(1, maxit + 1):
        dfx = df(x)
        if dfx == 0:
            raise ZeroDivisionError("Derivada nula; el metodo falla.")
        x_new = x - f(x) / dfx
        if abs(x_new - x) < tol:
            return x_new, i
        x = x_new
    return x, maxit


if __name__ == "__main__":
    f = lambda x: x**3 - x - 2
    df = lambda x: 3 * x**2 - 1
    raiz, n = newton_raphson(f, df, 1.5)
    print(f"Raiz ~= {raiz:.8f} en {n} iteraciones")
